Platform/OpenNtfsDxe/Compression.c

```diff
@@ -0,0 +1,668 @@
+    return Status;
+  }
+
+  *Result = (((UINT16) ByteHigh) << 8) + ByteLow;
+
+  return EFI_SUCCESS;
+}
+
+/**
+  * The basic idea is that substrings of the block which have been seen before
+    are compressed by referencing the string rather than mentioning it again.
+
+    #include <ntfs.h>\n
+    #include <stdio.h>\n
+
+    This is compressed to #include <ntfs.h>\n(-18,10)stdio(-17,4)
+
+  * Pairs like (-18,10) are recorded in two bytes. ->
+    The shortest possible substring is 3 bytes long. ->
+    One can subtract 3 from the length before encoding it.
+
+  * The references are always backward, and never 0. ->
+    One can store them as positive numbers, and subtract one. ->
+    (-18,10) -> (17,7); (-17,4) -> (16,1).
+
+  * Given that a block is 4096 in size, you might need 12 bits to encode the back reference.
+    This means that you have only 4 bits left to encode the length.
+
+  * Dynamic allocation of bits for the back-reference.
+
+    for (i = clear_pos - 1, lmask = 0xFFF, dshift = 12; i >= 0x10; i >>= 1) {
+      lmask >>= 1; // bit mask for length
+      dshift--;    // shift width for delta
+     }
+
+  * Now that we can encode a (offset,length) pair as 2 bytes,
+    we still have to know whether a token is back-reference, or plain-text.
+    This is 1 bit per token. 8 tokens are grouped together
+    and preceded with the tags byte.
+
+    ">\n(-18,10)stdio" would be encoded as "00000100"
+    (the 1 bit indicates the back reference).
+
+    "00000000"#include"00000000" <ntfs.h"00000100">\n(17,7)stdio"00000001"(16,1)
+
+  * As a compression unit consists of 16 clusters (default), it usually contains more than one of these blocks.
+    If you want to access the second block, it would be a waste of time to decompress the first one.
+    Instead, each block is preceded by a 2-byte length.
+    The lower 12 bits are the length, the higher 4 bits are of unknown purpose.
+    Actually, (n-1) is stored in the low 12 bits.
+
+  * The compression method is based on independently compressing blocks of X clusters,
+    where X = 2 ^ ATTR_HEADER_NONRES.CompressionUnitSize.
+
+  * If the block grows in size, it will be stored uncompressed.
+    A length of exactly 4095 is used to indicate this case.
+
+  * Bit 0x8000 is the flag specifying that the block is compressed.
+**/
+STATIC
+EFI_STATUS
+DecompressBlock (
+  IN  COMPRESSED *Clusters,
+  OUT UINT8      *Dest
+  )
+{
+  EFI_STATUS   Status;
+  UINT16       BlockParameters;
+  UINTN        BlockLength;
+  UINT8        TagsByte;
+  UINT8        Tokens;
+  UINT16       ClearTextPointer;
+  UINT8        PlainText;
+  UINT16       Index;
+  UINT16       Length;
+  UINT16       Lmask;
+  UINT16       Delta;
+  UINT16       Dshift;
+  UINT16       BackReference;
+  UINTN        SpareBytes;
+
+  Status = GetTwoDataRunBytes (Clusters, &BlockParameters);
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  BlockLength = (BlockParameters & BLOCK_LENGTH_BITS) + 1;
+
+  if (Dest != NULL) {
```

Comment:
Maybe inverse the condition here to make it shorter?

Reply:
In my opinion, the current order of instructions good enough correspond to the documentation above the function. As this function is a bit over complicated, such a local optimization will deteriorate that correspondence.